Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java

```diff
@@ -525,3 +577,117 @@ public String verbatimString() {
         return string;
     }
 }
+
+final class ConcatDafnySequence<T> extends DafnySequence<T> {
+    // INVARIANT: Either these are both non-null and ans is null or both are
+    // null and ans is non-null.
+    private DafnySequence<T> left, right;
+    // XXX We're storing the combined sequence as a DafnySequence, not an array,
+    // because it might need a backing array of primitive type.  Unfortunately,
+    // this means the resulting sequence is doomed to have a bit of overhead (an
+    // extra indirect function call) on every call to select().  Hard to see how
+    // to fix this without requiring that *every* DafnySequence suffer some
+    // overhead.
+    private DafnySequence<T> ans = null;
+    private final int length;
+
+    ConcatDafnySequence(DafnySequence<T> left, DafnySequence<T> right) {
```

Comment:
arguments can be `final`

Reply:
Should I just declare all the arguments to be `final`?  I wouldn't have made any arguments `final` except that I didn't know they'd loosened the restriction on variables being referenced from an inner class.

Reply:
It's just a style thing, since the arguments are "effectively final" anyway.  Since `final` doesn't mean `immutable` here, and the body is tiny, I withdraw the suggestion.  Generally I like making boxed arguments final, but it's a lot of typing for not much benefit.